diff --git a/include/sirit/sirit.h b/include/sirit/sirit.h
index 3d406de..b99c0c9 100644
--- a/include/sirit/sirit.h
+++ b/include/sirit/sirit.h
@@ -367,6 +367,9 @@ public:
         return OpStore(pointer, object, std::span<const MemoryAccessLiteral>({memory_access...}));
     }
 
+    /// Query the number of elements in the runtime array member of a struct pointer.
+    Id OpArrayLength(Id result_type, Id struct_pointer, std::uint32_t array_member);
+
     /// Create a pointer into a composite object that can be used with OpLoad and OpStore.
     Id OpAccessChain(Id result_type, Id base, std::span<const Id> indexes = {});
 
diff --git a/src/instructions/memory.cpp b/src/instructions/memory.cpp
index d6bdf0c..4dfa6e1 100644
--- a/src/instructions/memory.cpp
+++ b/src/instructions/memory.cpp
@@ -28,6 +28,12 @@ Id Module::OpStore(Id pointer, Id object, std::span<const MemoryAccessLiteral> m
     return *code << spv::Op::OpStore << pointer << object << memory_access << EndOp{};
 }
 
+Id Module::OpArrayLength(Id result_type, Id struct_pointer, std::uint32_t array_member) {
+    code->Reserve(5);
+    return *code << OpId{spv::Op::OpArrayLength, result_type} << struct_pointer << array_member
+                 << EndOp{};
+}
+
 Id Module::OpAccessChain(Id result_type, Id base, std::span<const Id> indexes) {
     assert(!indexes.empty());
     code->Reserve(4 + indexes.size());
